code/modules/recycling/conveyor2.dm

```diff
@@ -123,35 +125,43 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
 		return FALSE
 	return TRUE
 
-	// machine process
-	// move items to the target location
-/obj/machinery/conveyor/process()
-	if(machine_stat & (BROKEN | NOPOWER))
-		return
-	//If the conveyor is broken or already moving items
-	if(!operating || conveying)
-		return
-	use_power(6)
-	//get the first 30 items in contents
+// returns true if there is at least 1 item on the belt that can be moved
+/obj/machinery/conveyor/proc/check_belt_contents()
+	for(var/atom/movable/AM in loc)
+		if(!AM.anchored)
+			return TRUE
+	return FALSE
+
+// populates `affecting` with the first 30 items in the belt's contents
+/obj/machinery/conveyor/proc/get_belt_contents()
 	affecting = list()
 	var/i = 0
-	for(var/item in loc.contents)
-		if(item == src)
+	for(var/atom/movable/AM in loc)
+		if(AM.anchored) // If it can't be moved, ignore it
 			continue
-		i++ // we're sure it's a real target to move at this point
-		if(i >= MAX_CONVEYOR_ITEMS_MOVE)
+		if(++i > MAX_CONVEYOR_ITEMS_MOVE)
 			break
-		affecting.Add(item)
-	conveying = TRUE
-	addtimer(CALLBACK(src, .proc/convey, affecting), 1)
+		affecting.Add(AM)
 
+// when some item or mob moves over the conveyor belt
+/obj/machinery/conveyor/proc/conveyorCrossed(datum/source, atom/movable/AM, atom/oldLoc)
+	// if the conveyor isn't switched on or is broken / has no power
+	if(!operating || machine_stat & (BROKEN | NOPOWER))
+		return
+
+	get_belt_contents() // populate affecting with the movable items on the belt
```

Comment:
Should this be called every time a movable atom enters said turf?
It looks like this creates a lot of redundancy.
You might want to handle each entered atom by itself instead of triggering a `loc` contents check every time.

Reply:
It is probably better to do it on an atom by atom basis yeah. I think I kept this around because there's suppose to be a 30 item move limit and I didn't want to mess with that, but I can take another look.

Reply:
The problem here is that this will be triggered every time an item moves here, so it's potentially happeing a lot of times per second without checks, unlike before.

Reply:
Now I remember why I did it this way. This is to solve an issue where, if you for some reason stop over a conveyor belt that's running, say you run into some plastic flaps for example. And then you rest to continue getting pushed, the belt wouldn't push you, because the belt pushing is event based, only getting triggered when something moves over it.

The way I have it now, while not ideal, ensures that things sitting on belts and not moving, get moving when they are able. I'm not really sure what to do about this issue.

Reply:
The solution is to account for the event of the density (or anchored, or other conditions that prevent moving) changing.

Reply:
Well I messed around with adding more handling for these events, and had it all working fine. However when I profiled it compared to just using the standard `process()` instead, signal based conveying is actually more expensive.

I think I'll just revert back to use `process()` for conveyors, but still keep some of the code improvements I've made.

Reply:
It's probably more expensive but also more responsive, yes.